File: frontier/orchestrator/src/main/kotlin/dev/inprogress/frontier/ProbeWorkflow.kt
package dev.inprogress.frontier

import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.annotation.Workflow
import dev.restate.sdk.kotlin.runBlock
import dev.restate.sdk.kotlin.workflowKey

private const val CRASH_AFTER_EXECUTOR_ENV = "IN_PROGRESS_FRONTIER_CRASH_AFTER_EXECUTOR"

internal fun validatedProbeRequest(operationId: String, input: ProbeInput): ProbeRequest =
  try {
    ProbeRequest(
      operationId = ProbeContract.operationId(operationId),
      kind = ProbeContract.KIND,
      input = ProbeContract.input(input),
    ).also { ProbeContract.requestBody(it) }
  } catch (error: IllegalArgumentException) {
    throw TerminalException(TerminalException.BAD_REQUEST_CODE, error.message ?: "invalid probe request")
  }

@Workflow
class ProbeWorkflow(
  private val executor: ExecutorClient = ExecutorClient(),
  private val crashAfterExecutor: Boolean = System.getenv(CRASH_AFTER_EXECUTOR_ENV) == "1",
) {
  @Workflow
  suspend fun run(input: ProbeInput): ProbeResult {
    val request = validatedProbeRequest(workflowKey(), input)
    // Restate journals the response; the executor independently deduplicates the physical effect.
    return runBlock("idempotent executor call") {
      executor.execute(request).also {
        if (crashAfterExecutor) Runtime.getRuntime().halt(86)
      }
    }
  }
}

File: frontier/orchestrator/src/probe-workflow.ts
import * as restate from "@restatedev/restate-sdk";

import {
  PROBE_KIND,
  ProbeInputSchema,
  ProbeRequestSchema,
  probeRequestBody,
  type ProbeInput,
  type ProbeRequest,
  type ProbeResult,
} from "./contracts.ts";
import { ExecutorClient } from "./executor-client.ts";

export interface ProbeExecutor {
  execute(request: ProbeRequest): Promise<ProbeResult>;
}

export interface ProbeDurability {
  readonly key: string;
  run<T>(name: string, action: () => Promise<T>): PromiseLike<T>;
}

export function validatedProbeRequest(operationId: string, input: unknown): ProbeRequest {
  try {
    const request = ProbeRequestSchema.parse({
      operationId,
      kind: PROBE_KIND,
      input: ProbeInputSchema.parse(input),
    });
    probeRequestBody(request);
    return request;
  } catch (error) {
    const message = error instanceof Error ? error.message : "invalid probe request";
    throw new restate.TerminalError(message, { errorCode: 400 });
  }
}

export async function runProbe(
  ctx: ProbeDurability,
  input: ProbeInput,
  executor: ProbeExecutor,
  crashAfterExecutor?: () => never,
): Promise<ProbeResult> {
  const request = validatedProbeRequest(ctx.key, input);
  return await ctx.run("idempotent executor call", async () => {
    const result = await executor.execute(request);
    crashAfterExecutor?.();
    return result;
  });
}

const executor = new ExecutorClient();
const crashAfterExecutor =
  process.env.IN_PROGRESS_FRONTIER_CRASH_AFTER_EXECUTOR === "1"
    ? () => process.exit(86)
    : undefined;

export const probeWorkflow = restate.workflow({
  name: "ProbeWorkflow",
  handlers: {
    run: async (ctx: restate.WorkflowContext, input: ProbeInput) =>
      await runProbe(ctx, input, executor, crashAfterExecutor),
  },
});
